Token-based authentication 1: JWT + JWKS

zvonand opened the pull request:
First part of #89978 -- splitting that into parts would make it easier to review. 

### Changelog category (leave one):
- New Feature


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Support JWT-based authentication

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

zvonand committed:
Add token authentication and authorization

4ertus2 commented on src/Server/HTTP/authenticateUserByHTTP.cpp:
```diff
@@ -77,6 +78,8 @@ bool authenticateUserByHTTP(
     bool has_http_credentials = request.hasCredentials() && request.get("Authorization") != "never";
     bool has_credentials_in_query_params = params.has("user") || params.has("password");
 
+    std::string jwt_token = request.get("Authorization", (params.has("token") ? BEARER_PREFIX + params.get("token") : ""));
```
In #88857 "BEARER" could contain any token for future HTTPS external auth.

zvonand commented on src/Access/TokenProcessorsJWT.cpp:
```diff
@@ -0,0 +1,397 @@
+    if (!claims.empty() && !check_claims(claims, decoded_jwt.get_payload_json()))
+        return false;
+
+    credentials.setUserName(decoded_jwt.get_payload_claim(username_claim).as_string());
```
What do you mean? I did not quite understand how are you planning to match incoming token -> username.

Using some claim's value as a username is a common scenario. E.g. `sub` claim (which is used to extract username by default, can be changed in `config.xml`) is present in token in like 99% cases. Even if not -- you (as CH admin) know what service you are working with and you know what its token looks like, and you can configure it accordingly

zvonand replied:
A-ha, that's a very valid point. Using several claims makes sense indeed. I can add this rather easily. 

> It could be 'service_id' + 'service_operation' encoded in claims.

You mean, username would be combination of values in those fields in JWT, right?

zvonand commented on src/Server/HTTP/authenticateUserByHTTP.cpp:
```diff
@@ -212,6 +215,16 @@ bool authenticateUserByHTTP(
+        const auto token_credentials = TokenCredentials(jwt_token.substr(BEARER_PREFIX.length()));
+        const auto & external_authenticators = global_context->getAccessControl().getExternalAuthenticators();
+
+        if (!external_authenticators.checkTokenCredentials(token_credentials))
```
> It's not clean if we really allowed to check credentials here

Why? Here (and in other auth methods branches nearby) we simply form appropriate credentials based on incoming request. No logic like "allowed/not allowed to authenticate this user with this method" takes place here, it is done later. It cannot be done _before_ we processed incoming token and extracted username out of it.

Call to `external_authenticators.something()` may look inappropriate here, I agree. I will take a look into this and try to separate this logic even further, thanks.

zvonand commented on src/Access/TokenProcessorsJWT.cpp:
```diff
@@ -0,0 +1,397 @@
+namespace
+{
+
+bool check_claims(const picojson::value & claims, const picojson::value & payload, const String & path);
```
IDK if it is a good idea actually. They are now as simple as possible.

They are only used inside this file, and they use other CH code (i.e. logging). Now they are simple: they return bool (OK/not OK) and log the errors.

If I move them to `-cmake` dir, I will have to use exceptions (instead of logging) here or return tuples (which would make the code even worse and less readable)

zvonand committed:
small changes

zvonand committed:
small fix

4ertus2 commented on src/Parsers/CommonParsers.h:
```diff
@@ -84,6 +84,7 @@ namespace DB
     MR_MACROS(CHECK_TABLE, "CHECK TABLE") \
     MR_MACROS(CHECK_GRANT, "CHECK GRANT") \
     MR_MACROS(CHECK, "CHECK") \
+    MR_MACROS(CLAIMS, "CLAIMS") \
```
I think it should be an option. There's the most simple static JWK use cases where user passes JWK as some kind of hash he do not interpret.

Your case with required claims is interested too. I think both are valid.

zvonand committed:
defer token validation as far as possible

zvonand commented:
@bharatnc could you pls take a look / allow to run CI? This one or a bigger one: https://github.com/ClickHouse/ClickHouse/pull/89978

alexey-milovidov committed:
Delete aspell-dict.txt (removed in master)

`aspell-dict.txt` was deleted in master along with the `aspell` and
`codespell` style checks. Removing the file here so the PR can merge
cleanly without a modify/delete conflict.
